### PR TITLE
Add tests with invalid witnesses

### PR DESCRIPTION
**Pull request by georgwiese**
Some small changes and utility functions to make it easier to make tests that assert that proof generation fails if the witness is invalid.

I think we should add more in the future, also including lookups and permutation, but this one already needed a lot of changes, so I'll postpone that to another PR.

### CODE REVIEW

File: pipeline/src/test_util.rs [hunk 1]
```diff
@@ -18,32 +18,32 @@ pub fn resolve_test_file(file_name: &str) -> PathBuf {
...
     let pipeline = Pipeline::default()
         .from_asm_string(contents.to_string(), Some(PathBuf::from(file_name)))
         .with_prover_inputs(inputs)
         .add_external_witness_values(external_witness_values);
-    verify_pipeline(pipeline)
+    verify_pipeline(pipeline).unwrap();
 }
 
-pub fn verify_pipeline<T: FieldElement>(pipeline: Pipeline<T>) {
+pub fn verify_pipeline(pipeline: Pipeline<GoldilocksField>) -> Result<(), String> {
```
Comment (georgwiese) [inline pipeline/src/test_util.rs:46]: This function only does Pilcom verification, so I changed that it only accepts a Goldilocks pipeline.

File: pipeline/tests/pil.rs [hunk 2]
```diff
@@ -23,22 +24,14 @@ fn test_invalid_witness_pilcom() {
...
-        .from_file(resolve_test_file(f))
-        .set_witness(vec![(
-            "main.w".to_string(),
-            vec![GoldilocksField::from(0); 4],
-        )])
-        .with_backend(powdr_backend::BackendType::EStark)
-        .compute_proof()
-        .unwrap();
+    let witness = vec![("main.w".to_string(), vec![0; 4])];
+    assert_proofs_fail_for_invalid_witnesses(f, &witness);
```
Comment (georgwiese) [inline pipeline/tests/pil.rs:21]: This now tests 2 backend, Halo2 and eStark. It also asserts that Pilcom fails.

File: test_data/pil/fibonacci.pil [hunk 3]
```diff
@@ -1,4 +1,4 @@
-constant %N = 16;
+constant %N = 4;
```
Comment (georgwiese) [inline test_data/pil/fibonacci.pil:1]: To keep the tests small.

File: pipeline/tests/asm.rs [hunk 4]
```diff
@@ -2,8 +2,8 @@ use powdr_number::{FieldElement, GoldilocksField};
 use powdr_pipeline::test_util::{gen_estark_proof, test_halo2, verify_test_file};
 use test_log::test;
 
-fn verify_asm<T: FieldElement>(file_name: &str, inputs: Vec<T>) {
-    verify_test_file(file_name, inputs, vec![]);
+fn verify_asm(file_name: &str, inputs: Vec<GoldilocksField>) {
```
Comment (leonardoalt) [inline pipeline/tests/asm.rs:5]: feels actually odd we don't test BN here at all

File: pipeline/tests/pil.rs [hunk 5]
```diff
@@ -54,19 +35,6 @@ fn test_invalid_witness_halo2mock() {
...
 }
 
-#[test]
-#[should_panic = "called `Result::unwrap()` on an `Err` value: [\"Proof is invalid\"]"]
-#[cfg(feature = "halo2")]
-fn test_invalid_witness_halo2() {
-    let f = "pil/trivial.pil";
-    Pipeline::default()
-        .from_file(resolve_test_file(f))
-        .set_witness(vec![("main.w".to_string(), vec![Bn254Field::from(0); 4])])
```
Comment (leonardoalt) [inline pipeline/tests/pil.rs:64]: is this not tested anymore?

File: pipeline/src/test_util.rs [hunk 6]
```diff
@@ -18,32 +18,32 @@ pub fn resolve_test_file(file_name: &str) -> PathBuf {
...
 }
 
-pub fn verify_test_file<T: FieldElement>(
+pub fn verify_test_file(
     file_name: &str,
-    inputs: Vec<T>,
-    external_witness_values: Vec<(String, Vec<T>)>,
-) {
+    inputs: Vec<GoldilocksField>,
+    external_witness_values: Vec<(String, Vec<GoldilocksField>)>,
```
Comment (leonardoalt) [inline pipeline/src/test_util.rs:24]: again, feels weird that this is all GL only

File: pipeline/src/test_util.rs [hunk 7]
```diff
@@ -183,3 +183,67 @@ pub fn evaluate_integer_function<T: FieldElement>(
...
+#[cfg(not(feature = "halo2"))]
+fn assert_proofs_fail_for_invalid_witnesses_bn254(
+    _file_name: &str,
+    _witness: &[(String, Vec<u64>)],
+) {
+}
+
+pub fn assert_proofs_fail_for_invalid_witnesses(file_name: &str, witness: &[(String, Vec<u64>)]) {
+    assert_proofs_fail_for_invalid_witnesses_gl(file_name, witness);
+    assert_proofs_fail_for_invalid_witnesses_bn254(file_name, witness);
```
Comment (leonardoalt) [inline pipeline/src/test_util.rs:248]: where is it used with bn?

Comment (georgwiese) [reply pipeline/src/test_util.rs:248]: What do you mean?

File: pipeline/tests/pil.rs (hunk 5 above)
Comment (georgwiese) [reply pipeline/tests/pil.rs:64]: It is, `test_invalid_witness()` above calls `assert_proofs_fail_for_invalid_witnesses()` which tests Pilcom, eStark and Halo2 prover. The only thing it doesn't test is Halo2-Mock, which is why we keep the test above.

File: pipeline/tests/asm.rs (hunk 4 above)
Comment (georgwiese) [reply pipeline/tests/asm.rs:5]: I agree, but it was like that before.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**